examples/linear_model/plot_sgd_early_stopping.py

```diff
@@ -0,0 +1,155 @@
+until the validation score did not improve by at least ``tol`` during the last
+``n_iter_no_change`` iterations. The actual number of iterations is available
+at the attribute ``n_iter_``.
+
+This example illustrates how the early stopping can used in the
+:class:`sklearn.linear_model.SGDClassifier` model to achieve almost the same
+accuracy as compared to a model built without early stopping. This can
+significantly reduce training time. Note that scores differ between the
+stopping criteria even from early iterations because some of the training data
+is held out with the validation stopping criterion.
+"""
+# Authors: Tom Dupre la Tour
+#
+# License: BSD 3 clause
+from __future__ import print_function
+import time
+import sys
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn import linear_model
+from sklearn.datasets import fetch_mldata
+from sklearn.model_selection import train_test_split
+from sklearn.utils.testing import ignore_warnings
+from sklearn.exceptions import ConvergenceWarning
+from sklearn.utils import shuffle
+
+print(__doc__)
+
+
+def load_mnist(n_samples=None, class_0=0, class_1=8):
+    """Load MNIST, select two classes, shuffle and return only n_samples."""
+    mnist = fetch_mldata('MNIST original')
+
+    # take only two classes for binary classification
+    mask = np.logical_or(mnist.target == class_0, mnist.target == class_1)
+
+    X, y = shuffle(mnist.data[mask], mnist.target[mask], random_state=42)
+    if n_samples is not None:
+        X, y = X[:n_samples], y[:n_samples]
+    return X, y
+
+
+@ignore_warnings(category=ConvergenceWarning)
+def fit_and_score(estimator, max_iter, X_train, X_test, y_train, y_test):
+    """Fit the estimator on the train set and score it on both sets"""
+    estimator.set_params(max_iter=max_iter)
+
+    start = time.time()
+    estimator.fit(X_train, y_train)
+
+    fit_time = time.time() - start
+    n_iter = estimator.n_iter_
+    train_score = estimator.score(X_train, y_train)
+    test_score = estimator.score(X_test, y_test)
+
+    return fit_time, n_iter, train_score, test_score
+
+
+# Define the estimators to compare
+estimator_dict = {
+    'No stopping criterion':
+    linear_model.SGDClassifier(tol=None, random_state=0),
+    'Training loss':
+    linear_model.SGDClassifier(early_stopping=False, n_iter_no_change=5,
+                               tol=0.01, random_state=0),
+    'Validation score':
+    linear_model.SGDClassifier(early_stopping=True, n_iter_no_change=5,
+                               tol=0.01, random_state=0,
+                               validation_fraction=0.2)
+}
+
+# Load the dataset
+X, y = load_mnist(n_samples=10000)
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5,
+                                                    random_state=0)
+
+max_iter = 50
+results = []
+for estimator_name, estimator in estimator_dict.items():
+    print(estimator_name + ': ', end='')
+    for this_max_iter in range(1, max_iter):
```

Comment:
instead of fitting 50 model for each criterio I would rather do:

```python
for estimator_name, estimator in estimator_dict.items():
    print(estimator_name + ': ', end='') 
    for random_state in range(5):
        for this_max_iter in [1, 5, 10, 15, 20, 30]:
            print('.', end='')
            sys.stdout.flush()

            estimator.set_params(random_state=random_state)
            fit_time, n_iter, train_score, test_score = fit_and_score(
                estimator, this_max_iter, X_train, X_test, y_train, y_test)

            results.append((estimator_name, this_max_iter, random_state,
                            fit_time, n_iter, train_score, test_score))
     print('')

Reply:
But then we don't see the oscillations of the train and test score.

Reply:
Alright. That's a good point.